### PR TITLE
Adding example for Cloud Dataproc

### PR DESCRIPTION
**Pull request by evilsoapbox**
Go example for Google Cloud Dataproc. This sample follows a pretty basic (and common) Cloud Dataproc pattern:
1. Create cluster
2. Submit job
3. Get job output
4. Delete cluster

A sample PySpark job is also included in the repository.

### CODE REVIEW

**Comment by broady** [comment]
High level comments:

1. Is there a better name for this sample? We're trying to avoid putting samples at the top level directory: it makes it more difficult to add more samples for the same product again in the future. See #75 for a recent example.
2. Please use the `cloud.google.com/go/storage` package for GCS.
3. Package level doc should be in the form of `Command foo ...` or `Sample foo ...`
4. Please use the license header found in all the other files.
5. Function-level comments should be in this form:
```
CreateCluster creates a Cloud Dataproc cluster with the given name, region, project and zone.
func CreateCluster(...
```
6. Thanks for writing a test, but the test provided doesn't actually test anything. See the tests in #75 for something more appropriate for this kind of sample.
7. Only the `main` function should call `log.Fatal`.
8. Please put the `main` function at the top of the file. It's difficult to recognize the program flow otherwise.
9. Error variables should be named `err` unless there's a good reason for them to be named something else.
10. I believe we use the term `region`, not `regionID` (across Cloud).

**Comment by evilsoapbox** [comment]
1: How about putting it under something like `/data/dataproc/...` which would also allow other GCP data products, like Cloud Dataflow, to use the base directory `/data/dataflow/...` or BigQuery `/data/bigquery/...`. LMK what you think and I will reorganize.
2: Fixed
3: Fixed
4: Fixed
5: Fixed
6: Fixed
7: Fixed
8: Fixed
9: Fixed
10: Fixed

I just pushed an update to my fork based on the feedback.

As a note, in the unit tests the `TestJobFunctionality()` method is pretty weighty. The thinking here is all ops are dependent on one another and not well run atomically. Let me know if it's too weighty.

File: dataproc/dataproc.go [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+		log.Fatal(err)
+	}
+	fmt.Println("Cluster deleted")
+	fmt.Printf("Done...")
+}
+
+// getDataprocService creates and returns a Cloud Dataproc service object.
+func getDataprocService() (service *dataproc.Service, err error) {
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, scope)
```
Comment (broady) [inline dataproc/dataproc.go]: This error is ignored. I suggest inlining this (move all of this up to `main()`) too.

File: dataproc/dataproc.go [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+	// Create a new Dataproc service
+	service, err = dataproc.New(client)
+
+	return service, err
+}
+
+// getStorageService creates and returns a Cloud Storage service object.
+func getStorageService() (service *storage.Client, err error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
```
Comment (broady) [inline dataproc/dataproc.go]: Same here.

File: dataproc/dataproc.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+	// Create a new storage service
+	storageService, err := getStorageService()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Create a cluster
+	_, err = createCluster(service, *clusterName, *region, *projectID, *zoneID)
+	if err != nil {
```
Comment (broady) [inline dataproc/dataproc.go]: `if _, err := createCluster(...); err != nil {`

File: dataproc/dataproc.go [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+		log.Fatal(err)
+	}
+	fmt.Println("Cluster created")
```
Comment (broady) [inline dataproc/dataproc.go]: log.Print?

File: dataproc/dataproc.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+	// Get the job output
+	clusterData, err := getClusterDataByName(service, *clusterName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	output, err := getJobOutput(storageService, *projectID, clusterData[1], clusterData[3], jobId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Printf("Job output:\n%s\n", string(output))
```
Comment (broady) [inline dataproc/dataproc.go]: No need for `string()` - you're already using `%s` in the format string.

File: dataproc/dataproc.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+		log.Fatal(err)
+	}
+	fmt.Println("Job is finished")
+
+	// Get the job output
+	clusterData, err := getClusterDataByName(service, *clusterName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	output, err := getJobOutput(storageService, *projectID, clusterData[1], clusterData[3], jobId)
```
Comment (broady) [inline dataproc/dataproc.go]: what are 1 and 3 here?

File: dataproc/dataproc.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+	}
+	fmt.Printf("Job output:\n%s\n", string(output))
+
+	// Delete the cluster
+	_, err = deleteCluster(service, *projectID, *region, *clusterName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println("Cluster deleted")
+	fmt.Printf("Done...")
```
Comment (broady) [inline dataproc/dataproc.go]: `Done..` is quite extraneous.

File: dataproc/dataproc.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+func getStorageService() (service *storage.Client, err error) {
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+
+	return client, err
+}
+
+// createCluster creates a Cloud Dataproc cluster with the given name and region.
+func createCluster(service *dataproc.Service, name string, region string,
+	project string, zone string) (response string, err error) {
```
Comment (broady) [inline dataproc/dataproc.go]: Wrapping lines is a good sign that the parameters should go into a struct. These get passed around an awful lot.

Would something like this make sense?

```
type clusterConfig struct{
    project, region, zone, name string
}
```

File: dataproc/dataproc.go [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+// getClusterDataByName returns metadata about the cluster with the given name.
+func getClusterDataByName(service *dataproc.Service, name string) (clusterData []string, err error) {
+	// Get a list of clusters
+	clusters, err := listClusters(service)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the cluster requested in the list
+	for _, item := range clusters {
```
Comment (broady) [inline dataproc/dataproc.go]: `for _, c := range clusters`

File: dataproc/dataproc.go [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+		if item[0] == name {
+			return item, err
+		}
+	}
+
+	return nil, errors.New("Cluster not found")
```
Comment (broady) [inline dataproc/dataproc.go]: Error strings should start with a lowercase letter. They are often concatenated with a prefix - having a capital in the middle is ugly.

File: dataproc/dataproc.go [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+// getJobOutput returns the text from the job (raw driver output) with the given project, cluser name, bucket id, and job id.
+func getJobOutput(storageClient *storage.Client, project string, cluster string, bucket string, job string) (output []byte, err error) {
+
+	// Format the object name
+	object := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", cluster, job)
+
+	// Read the file
+	rc, err := storageClient.Bucket(bucket).Object(object).NewReader(context.Background())
```
Comment (broady) [inline dataproc/dataproc.go]: `r, err := `

File: dataproc/dataproc.go [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	return nil, errors.New("Cluster not found")
+}
+
+// getJobOutput returns the text from the job (raw driver output) with the given project, cluser name, bucket id, and job id.
+func getJobOutput(storageClient *storage.Client, project string, cluster string, bucket string, job string) (output []byte, err error) {
+
+	// Format the object name
+	object := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", cluster, job)
```
Comment (broady) [inline dataproc/dataproc.go]: Would be good to have a reference link here as a comment - where does this string format come from?

File: dataproc/dataproc.go [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+
+	return res.Status.State, err
+}
+
+// listClusters lists all clusters in the current project.
+func listClusters(service *dataproc.Service) (clusters [][]string, err error) {
+	// List all clusters in a project for a given region
+	res, err := service.Projects.Regions.Clusters.List(*projectID, *region).Do()
+
+	for _, item := range res.Clusters {
```
Comment (broady) [inline dataproc/dataproc.go]: `for _, c := range ...`

File: dataproc/dataproc.go [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+	//Get the Job's status
+	res, err := service.Projects.Regions.Jobs.Get(project, region, jobId).Do()
+
+	return res.Status.State, err
+}
+
+// listClusters lists all clusters in the current project.
+func listClusters(service *dataproc.Service) (clusters [][]string, err error) {
+	// List all clusters in a project for a given region
+	res, err := service.Projects.Regions.Clusters.List(*projectID, *region).Do()
```
Comment (broady) [inline dataproc/dataproc.go]: Check this error.

File: dataproc/dataproc.go [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+
+	return output, err
+}
+
+// getJobStatus returns the status of the job with the given job id, project, and region.
+func getJobStatus(service *dataproc.Service, jobId string, project string, region string) (status string, err error) {
+	//Get the Job's status
+	res, err := service.Projects.Regions.Jobs.Get(project, region, jobId).Do()
+
+	return res.Status.State, err
```
Comment (broady) [inline dataproc/dataproc.go]: I'm sure this will result in a nil pointer panic.

Check the error.

File: dataproc/dataproc.go [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+	rc, err := storageClient.Bucket(bucket).Object(object).NewReader(context.Background())
+	output, err = ioutil.ReadAll(rc)
+	rc.Close()
+
+	return output, err
+}
+
+// getJobStatus returns the status of the job with the given job id, project, and region.
+func getJobStatus(service *dataproc.Service, jobId string, project string, region string) (status string, err error) {
+	//Get the Job's status
```
Comment (broady) [inline dataproc/dataproc.go]: Add a space after `//`

File: dataproc/dataproc.go [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+	// Upload the file to GCS
+	job_filename_parts := strings.Split(filepath, "/")
+	filename := job_filename_parts[len(job_filename_parts)-1]
+
+	ctx := context.Background()
+	wc := storageClient.Bucket(bucket).Object(filename).NewWriter(ctx)
+	wc.ContentType = "text/plain"
+	wc.ACL = []storage.ACLRule{{storage.AllUsers, storage.RoleReader}}
+	file, err := ioutil.ReadFile(filepath)
+	if err != nil {
```
Comment (broady) [inline dataproc/dataproc.go]: Read the file first - then you don't have the racy `os.Stat` call

File: dataproc/dataproc.go [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+			item.Config.ConfigBucket,
+		}
+		clusters = append(clusters, clusterDetails)
+	}
+
+	return clusters, err
+}
+
+// submitJob submits a PySpark job with the given file path to a PySpark file, project, bucket, and cluster.
+func submitJob(service *dataproc.Service, storageClient *storage.Client, filepath string, project string, bucket string, cluster string) (jobId string, err error) {
```
Comment (broady) [inline dataproc/dataproc.go]: `jobId` -> `jobID`

File: dataproc/dataproc.go (hunk 14 above)
Comment (broady) [inline dataproc/dataproc.go]: Also inconsistent use of flags vs function parameters (`*service, *region`)

File: dataproc/dataproc.go [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+	job := dataproc.Job{
+		Placement:  &placement,
+		PysparkJob: &pySparkJob,
+		Reference:  &jobReference,
+	}
+	jobRequest := dataproc.SubmitJobRequest{
+		Job: &job,
+	}
+
+	_, err = service.Projects.Regions.Jobs.Submit(*projectID, "global", &jobRequest).Do()
```
Comment (broady) [inline dataproc/dataproc.go]: Use `project` from the parameters.

File: dataproc/dataproc.go [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	// Submit the PySpark job
+	placement := dataproc.JobPlacement{
+		ClusterName: cluster,
+	}
+	pySparkJob := dataproc.PySparkJob{
+		MainPythonFileUri: "gs://" + bucket + "/" + filename,
+	}
+	jobID := "test-job-" + fmt.Sprintf("%v", time.Now().Unix())
```
Comment (broady) [inline dataproc/dataproc.go]: Why's it a `test-job`?

File: dataproc/dataproc.go [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+		}
+		if jobStatus == "DONE" {
+			finished = true
+		} else if jobStatus == "ERROR" {
+			finished = true
+			err = errors.New("Job finished with an error")
+		}
+
+		// Sleep for one second
+		time.Sleep(1000 * time.Millisecond)
```
Comment (broady) [inline dataproc/dataproc.go]: `time.Second`, remove comment.

File: dataproc/dataproc.go [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+
+// waitForJob waits for a job to finish with the given job id, project, and region.
+func waitForJob(service *dataproc.Service, jobId string, project string, region string) (finished bool, err error) {
+	for finished == false {
+		jobStatus, err := getJobStatus(service, jobId, project, region)
+		if err != nil {
+			return false, err
+		}
+		if jobStatus == "DONE" {
+			finished = true
```
Comment (broady) [inline dataproc/dataproc.go]: return true, nil

File: dataproc/dataproc.go [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+		} else if jobStatus == "ERROR" {
+			finished = true
+			err = errors.New("Job finished with an error")
```
Comment (broady) [inline dataproc/dataproc.go]: return true, errors.New("job errored")

File: dataproc/dataproc.go [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+
+		// Sleep for one second
+		time.Sleep(1000 * time.Millisecond)
+	}
+	return running, err
+}
+
+// waitForJob waits for a job to finish with the given job id, project, and region.
+func waitForJob(service *dataproc.Service, jobId string, project string, region string) (finished bool, err error) {
+	for finished == false {
```
Comment (broady) [inline dataproc/dataproc.go]: just
```
for {
   // ...
}
```

(use early returns)

**Comment by broady** [comment]
@evilsoapbox thanks!

Does `golang-samples/dataflow/runjob` make sense? I like the idea of keeping bigquery top-level, so dataflow probably should be, too.

I'll look at the tests again later - probably tomorrow.

File: dataproc/dataproc.go (hunk 16 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed.

File: dataproc/dataproc.go (hunk 3 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Good catch. Fixed.

File: dataproc/dataproc.go (hunk 4 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Yeah. That makes more sense. Fixed.

File: dataproc/dataproc.go (hunk 21 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 6 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Yeah, I should be explicit. They are UUIDs for a cluster and bucket. I have added an explicit assignment so someone reading this knows what they are. (If this should be a Map, I can do that too.)
```
	var clusterUuid = clusterData[1]
	var bucketUuid = clusterData[3]
	if output, err := getJobOutput(storageService, *projectID, clusterUuid, bucketUuid, jobId); err != nil {
```

File: dataproc/dataproc.go (hunk 5 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed.

File: dataproc/dataproc.go (hunk 7 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Removed.

File: dataproc/dataproc.go (hunk 1 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 2 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 22 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 23 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 24 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Sweet, done.

File: dataproc/dataproc.go (hunk 8 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Implemented throughout.

File: dataproc/dataproc.go (hunk 9 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed.

File: dataproc/dataproc.go (hunk 10 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Ack. Fixed.

File: dataproc/dataproc.go (hunk 12 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Ack. Fixed.

File: dataproc/dataproc.go (hunk 19 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed with struct.

File: dataproc/dataproc.go (hunk 20 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fair point. I removed the prefix so the job name is the unix time.

File: dataproc/dataproc.go (hunk 17 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Ack. Fixed.

File: dataproc/dataproc.go (hunk 18 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Good catch. Fixed.

File: dataproc/dataproc.go (hunk 15 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed.

File: dataproc/dataproc.go (hunk 14 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed.

**Comment by evilsoapbox** [comment]
@broady  Thanks for taking a look!

Just added a new commit with fixes and revised tests.

How about `golang-samples/dataproc/runjob`? Dataflow is a seperate GCP product and they may want to use that name. :)

File: dataproc/dataproc.go [hunk 25]
```diff
@@ -0,0 +1,304 @@
...
+
+const (
+	// Filename of example job (Python) file uploaded to Cloud Storage
+	exampleJobName = "test-file"
+
+	// Allows for full access to Google Cloud Platform products
+	scope = "https://www.googleapis.com/auth/cloud-platform"
+
+	// Compute URI base
+	computeUriBase = "https://www.googleapis.com/compute/v1/projects/%s/zones/%s"
```
Comment (broady) [inline dataproc/dataproc.go]: `s/Uri/URI`

Also, it's not really a base. It's a format string.

`computeURIFormat = ...`?

File: dataproc/dataproc.go [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+// clusterConfig defines the confuration of a cluster including its project, name, and region.
+type clusterConfig struct{
+		bucket,
+    project,
+		region,
+		state,
+		name,
+		uuid,
+		zone string
+}
```
Comment (broady) [inline dataproc/dataproc.go]: Run `gofmt -s .` from the root directory.

File: dataproc/dataproc.go [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+	client, err := google.DefaultClient(ctx, scope)
+	if err != nil {
+		log.Fatal(err)
+	}
+	service, err := dataproc.New(client)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Create a new Google Clout Storage service
```
Comment (broady) [inline dataproc/dataproc.go]: ouch!

Either `s/Clout/Cloud`, or better: remove these comments. It's clear that it's creating a new X service.

File: dataproc/dataproc.go [hunk 28]
```diff
@@ -0,0 +1,304 @@
...
+	storageService, err := storage.NewClient(ctx)
```
Comment (broady) [inline dataproc/dataproc.go]: `storageClient`

File: dataproc/dataproc.go [hunk 29]
```diff
@@ -0,0 +1,304 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Create a new clusterConfig to hold details about this cluster
+	cluster := clusterConfig{project: *projectID, region: *region, name: *clusterName, zone: *zoneID}
+
+	// Create a cluster
+	if _, err := createCluster(service, &cluster); err != nil {
```
Comment (broady) [inline dataproc/dataproc.go]: Add ctx as a first parameter to all of these.

File: dataproc/dataproc.go [hunk 30]
```diff
@@ -0,0 +1,304 @@
...
+		log.Fatal(err)
+	}
+	log.Println("Cluster created")
+
+	// Wait on the cluster to become active
+	log.Println("Waiting for cluster to be ready")
```
Comment (broady) [inline dataproc/dataproc.go]: Again, the comment says the same thing as the log statement. Remove the comments.

File: dataproc/dataproc.go [hunk 31]
```diff
@@ -0,0 +1,304 @@
...
+	log.Println("Job is finished")
+
+	output, err := getJobOutput(storageService, jobID, &cluster)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("Job output:\n%s\n", output)
+
+	// Delete the cluster
+	if _, err = deleteCluster(service, &cluster); err != nil {
```
Comment (broady) [inline dataproc/dataproc.go]: Change `err =` to `err :=` (reduces scope)

File: dataproc/dataproc.go [hunk 32]
```diff
@@ -0,0 +1,304 @@
...
+	clusterSpec := dataproc.Cluster{
+		ClusterName: cluster.name,
+		ProjectId:   cluster.project,
+		Config:      &clusterConfig,
+	}
+
+	// Create the cluster
+	res, err := service.Projects.Regions.Clusters.Create(cluster.project, cluster.region, &clusterSpec).Do()
+
+	return fmt.Sprintf("%s", res), err
```
Comment (broady) [inline dataproc/dataproc.go]: Just `return ...Do()`, perhaps? The formatted string is not used except for tests.

File: dataproc/dataproc.go [hunk 33]
```diff
@@ -0,0 +1,304 @@
...
+func deleteCluster(service *dataproc.Service, cluster *clusterConfig) (response string, err error) {
+	// Delete the cluster
+	res, err := service.Projects.Regions.Clusters.Delete(cluster.project, cluster.region, cluster.name).Do()
+
+	return fmt.Sprintf("%s", res), err
+}
+
+// getJobOutput returns the text from the job (raw driver output) with the given project, cluser name, bucket id, and job id.
+func getJobOutput(storageClient *storage.Client, job string, cluster *clusterConfig) (output []byte, err error) {
+
```
Comment (broady) [inline dataproc/dataproc.go]: Remove blank line.

File: dataproc/dataproc.go [hunk 34]
```diff
@@ -0,0 +1,304 @@
...
+	// Create the cluster
+	res, err := service.Projects.Regions.Clusters.Create(cluster.project, cluster.region, &clusterSpec).Do()
+
+	return fmt.Sprintf("%s", res), err
+}
+
+// deleteCluster deletes the Cloud Dataproc cluster with the given project, region, and name.
+func deleteCluster(service *dataproc.Service, cluster *clusterConfig) (response string, err error) {
+	// Delete the cluster
```
Comment (broady) [inline dataproc/dataproc.go]: Remove this comment

File: dataproc/dataproc.go (hunk 29 above)
Comment (broady) [inline dataproc/dataproc.go]: On all of these. No need to take the pointer to the cluster variable. Just pass the struct through.

File: dataproc/dataproc.go [hunk 35]
```diff
@@ -0,0 +1,304 @@
...
+	// Get the Job's status
+	res, err := service.Projects.Regions.Jobs.Get(project, region, jobID).Do()
+	if err != nil {
+		return "", err
+	}
+	return res.Status.State, nil
+}
+
+// listClusters lists all clusters in the current project.
+func listClusters(service *dataproc.Service, project string, region string) (clusters [][]string, err error) {
```
Comment (broady) [inline dataproc/dataproc.go]: A struct would be nicer than a `[][]string`

File: dataproc/dataproc.go [hunk 36]
```diff
@@ -0,0 +1,304 @@
...
+// submitJob submits a PySpark job with the given file path to a PySpark file, project, bucket, and cluster.
+func submitJob(service *dataproc.Service, storageClient *storage.Client, filepath string, bucket string, cluster *clusterConfig) (jobID string, err error) {
+	// Read the file from disk
+	file, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return "", err
+	}
+
+	// Upload the file to GCS
+	job_filename_parts := strings.Split(filepath, "/")
```
Comment (broady) [inline dataproc/dataproc.go]: Camel case, please

File: dataproc/dataproc.go [hunk 37]
```diff
@@ -0,0 +1,304 @@
...
+	jobRequest := dataproc.SubmitJobRequest{
+		Job: &job,
+	}
+
+	_, err = service.Projects.Regions.Jobs.Submit(cluster.project, cluster.region, &jobRequest).Do()
+	return jobID, err
+}
+
+// updateClusterMetadata geta metadata and updates the given clusterConfig.
+func updateClusterMetadata(service *dataproc.Service, cluster *clusterConfig) (err error) {
```
Comment (broady) [inline dataproc/dataproc.go]: Ah, yes, this one will need a pointer.

File: dataproc/dataproc.go [hunk 38]
```diff
@@ -0,0 +1,304 @@
...
+
+// clusterConfig defines the confuration of a cluster including its project, name, and region.
+type clusterConfig struct{
+		bucket,
+    project,
+		region,
+		state,
+		name,
+		uuid,
+		zone string
```
Comment (broady) [inline dataproc/dataproc.go]: A few of these fields are only set by the server, like uuid. I think only the user-set fields should be in a struct named `XConfig`

**Comment by broady** [comment]
This is looking very good. Thank you.

File: dataproc/dataproc.go (hunk 25 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Good point. Added some clarification and a link to the docs.

File: dataproc/dataproc.go (hunk 26 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 27 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done!

File: dataproc/dataproc.go (hunk 28 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done

File: dataproc/dataproc.go (hunk 36 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed.

File: dataproc/dataproc.go (hunk 34 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done

File: dataproc/dataproc.go (hunk 33 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 30 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 31 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Done.

File: dataproc/dataproc.go (hunk 32 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Removed the string output and just return the error now.

File: dataproc/dataproc.go (hunk 35 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Refactored to introduce a second struct `clusterDetails` to contain details about a created cluster. This cleanly (I think) separates a `config` which is used to create a cluster with the actual returned details from a cluster.

Unfortuinately, our API does not cleanly return `region` or `zone` so I have to parse some URLs to get them.

File: dataproc/dataproc.go (hunk 37 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed with the new struct.

File: dataproc/dataproc.go (hunk 38 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Fixed with the new struct.

File: dataproc/dataproc.go (hunk 29 above)
Comment (evilsoapbox) [inline dataproc/dataproc.go]: Updated to point to the struct without pointers.

With respect to the `ctx` to you mean line 86 in the `createCluster` and other methods? I'm new to Go and am looking for a design pattern around it but can't find a great example so I am a bit stuck there. :)

**Comment by evilsoapbox** [comment]
Thank you @broady for all of the feedback (and sorry for being a pain - coming to go from Python.) :) Just pushed a new commit with fixes.

**Comment by broady** [comment]
Thanks, @evilsoapbox. No worries! I'll probably take over the CL and do the final clean up and merge.

Gimme a few days at least. A bit backed up right now.

**Comment by shimont** [comment]
This example really helps!
Why dont you merge it?

**Comment by odeke-em** [comment]
How's it going @broady? Just a gentle ping here.

**Comment by tbpg** [comment]
Hi Folks. Sorry this PR fell off of the radar. Would you like to move forward with it?

I'm going to close it for now (it's been ~2 years since the last update). But, if we still need/want it and it will be embedded in documentation somewhere, we can reopen, clean it up, and move forward.

cc @bradmiro

**Pull request closed without merging**